Mathlib/RingTheory/PowerSeries/Order.lean

```diff
@@ -303,30 +311,58 @@ theorem order_X_pow (n : ℕ) : order ((X : R⟦X⟧) ^ n) = n := by
   rw [X_pow_eq, order_monomial_of_ne_zero]
   exact one_ne_zero
 
+@[simp]
+theorem divided_by_X_pow_order_of_X_eq_one :
+    divided_by_X_pow_order X_ne_zero = (1 : R⟦X⟧) := by
+  apply X_pow_mul_cancel
+  rw [self_eq_X_pow_order_mul_divided_by_X_pow_order]
+  simp
+
 end OrderZeroNeOne
 
-section OrderIsDomain
+section NoZeroDivisors
 
--- TODO: generalize to `[Semiring R] [NoZeroDivisors R]`
-variable [CommRing R] [IsDomain R]
+variable [Semiring R] [NoZeroDivisors R]
 
 /-- The order of the product of two formal power series over an integral domain
  is the sum of their orders. -/
 theorem order_mul (φ ψ : R⟦X⟧) : order (φ * ψ) = order φ + order ψ := by
-  classical
-  simp only [order_eq_emultiplicity_X]
-  exact emultiplicity_mul X_prime
-
--- Dividing `X` by the maximal power of `X` dividing it leaves `1`.
-@[simp]
-theorem divided_by_X_pow_order_of_X_eq_one : divided_by_X_pow_order X_ne_zero = (1 : R⟦X⟧) := by
-  rw [← mul_eq_left₀ X_ne_zero]
-  simpa using self_eq_X_pow_order_mul_divided_by_X_pow_order (@X_ne_zero R _ _)
+  apply le_antisymm _ (le_order_mul _ _)
+  by_cases h : φ.order = ⊤ ∨ ψ.order = ⊤
+  · rcases h with h | h <;> simp [h]
+  · simp only [not_or, ENat.ne_top_iff_exists] at h
+    obtain ⟨m, hm⟩ := h.1
+    obtain ⟨n, hn⟩ := h.2
+    rw [← hm, ← hn, ← ENat.coe_add]
+    rw [eq_comm, order_eq_nat] at hm hn
+    apply order_le
+    rw [coeff_mul, Finset.sum_eq_single ⟨m, n⟩]
+    · exact mul_ne_zero_iff.mpr ⟨hm.1, hn.1⟩
+    · intro ij hij h
+      rcases trichotomy_of_add_eq_add (mem_antidiagonal.mp hij) with h' | h' | h'
+      · exact False.elim (h (by simp [Prod.ext_iff, h'.1, h'.2]))
+      · rw [hm.2 ij.1 h', zero_mul]
+      · rw [hn.2 ij.2 h', mul_zero]
+    · intro h
+      apply False.elim (h _)
+      simp [mem_antidiagonal]
+
```

Comment:
This lemma below deserves being mentioned in the doc at the beginning, I guess. Also, do we want it to be a `simp` lemma? What do you think @riccardobrasca ?

Reply:
It is implicitely there (I only weaken the assumptions), and please don't ask me to redo the docstring that was the work of somebody else here.